nexus/db-queries/src/db/datastore/disk.rs

```diff
@@ -771,117 +787,287 @@ impl DataStore {
+                                    "cannot attach disk: instance is not \
+                                        fully stopped",
+                                ));
                             }
                             match collection.runtime_state.nexus_state.state() {
                                 // Ok-to-be-attached instance states:
-                                api::external::InstanceState::Creating |
-                                api::external::InstanceState::Stopped => {
+                                api::external::InstanceState::Creating
+                                | api::external::InstanceState::Stopped => {
                                     // The disk is ready to be attached, and the
                                     // instance is ready to be attached. Perhaps
                                     // we are at attachment capacity?
                                     if attached_count == i64::from(max_disks) {
-                                        return Err(Error::invalid_request(&format!(
-                                            "cannot attach more than {} disks to instance",
-                                            max_disks
-                                        )));
+                                        return Err(Error::invalid_request(
+                                            &format!(
+                                                "cannot attach more than \
+                                                {max_disks} disks to instance",
+                                            ),
+                                        ));
+                                    }
+
+                                    // Was this attach of a local storage disk
+                                    // blocked due to an existing sled resource
+                                    // record?
+                                    if local_storage_disk {
+                                        use sled_resource_vmm::dsl;
+
+                                        let record = dsl::sled_resource_vmm
+                                            .filter(
+                                                dsl::instance_id
+                                                    .eq(authz_instance.id()),
+                                            )
+                                            .select(dsl::instance_id)
+                                            .execute_async(&*conn)
+                                            .await
+                                            .optional()
+                                            .map_err(|e| {
+                                                public_error_from_diesel(
+                                                    e,
+                                                    ErrorHandler::Server,
+                                                )
+                                            })?;
```

Comment:
i was a bit surprised we didn't have a pre-existing query for looking up sled resource records for an instance, but i guess that table basically just gets read in the allocation and deallocation CTes so we haven't needed one for now. so, we probably don't need to think about factoring this out until we need to do it elsewhere :thumbsup:...